8253180: ZGC: Implementation of JEP 376: ZGC: Concurrent Thread-Stack Processing

fisk opened the pull request:
This PR the implementation of "JEP 376: ZGC: Concurrent Thread-Stack Processing" (cf. https://openjdk.java.net/jeps/376).

Basically, this patch modifies the epilog safepoint when returning from a frame (supporting interpreter frames, c1, c2, and native wrapper frames), to compare the stack pointer against a thread-local value. This turns return polls into more of a swiss army knife that can be used to poll for safepoints, handshakes, but also returns into not yet safe to expose frames, denoted by a "stack watermark".

ZGC will leave frames (and other thread oops) in a state of a mess in the GC checkpoint safepoints, rather than processing all threads and their stacks. Processing is initialized automagically when threads wake up for a safepoint, or get poked by a handshake or safepoint. Said initialization processes a few (3) frames and other thread oops. The rest - the bulk of the frame processing, is deferred until it is actually needed. It is needed when a frame is exposed to either 1) execution (returns or unwinding due to exception handling), or 2) stack walker APIs. A hook is then run to go and finish the lazy processing of frames.

Mutator and GC threads can compete for processing. The processing is therefore performed under a per-thread lock. Note that disarming of the poll word (that the returns are comparing against) is only performed by the thread itself. So sliding the watermark up will require one runtime call for a thread to note that nothing needs to be done, and then update the poll word accordingly. Downgrading the poll word concurrently by other threads was simply not worth the complexity it brought (and is only possible on TSO machines). So left that one out.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ❌ (2/5 failed) | ⏳ (2/2 running) | ✔️ (2/2 passed) |
| Build / test | ⏳ (1/1 running) |    |     | 
| Test (tier1) |    |     |  ⏳ (8/9 running) |

**Failed test tasks**
- [Linux x64 (build hotspot minimal)](https://github.com/fisk/jdk/runs/1230171906)
- [Linux x64 (build hotspot zero)](https://github.com/fisk/jdk/runs/1230171889)

### Issue
 * [JDK-8253180](https://bugs.openjdk.java.net/browse/JDK-8253180): ZGC: Implementation of JEP 376: ZGC: Concurrent Thread-Stack Processing


### Reviewers
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to 71db03039ef4f5d0c7cfcecae8e46a9c18570e03
 * [Per Liden](https://openjdk.java.net/census#pliden) (@pliden - **Reviewer**) ⚠️ Review applies to 4e2aabbb71c01bb6d7f60090b2ec60d46a1b8c7a
 * [Robbin Ehn](https://openjdk.java.net/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to 2ffbd764d8d9baa93f91ff9a8072ce576726bb6b
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to 3db0dbcb494b5639e3657fe09b42c7ef30136fab
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 078a207e2128f88fcb2f918a2e574bf2e202005f
 * [Stuart Monteith](https://openjdk.java.net/census#smonteith) (@stooart-mon - Author) ⚠️ Review applies to 54fe1f8b12f156ea4b29eaa28734fc348ef40fc5


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/296/head:pull/296`
`$ git checkout pull/296`

fisk commented:
/label hotspot

fisk commented:
/label hotspot-gc

fisk commented:
/label remove hotspot-gc

stefank reviewed (commented):
I've pre-reviewed most of the changes.

stefank commented on src/hotspot/cpu/x86/templateInterpreterGenerator_x86_64.cpp:
```diff
@@ -191,7 +191,7 @@ address TemplateInterpreterGenerator::generate_CRC32_update_entry() {
 
     Label slow_path;
-    __ safepoint_poll(slow_path, r15_thread, rscratch1);
+    __ safepoint_poll(slow_path, r15_thread, true /* at_return */, false /* in_nmethod */);
```
Why is this tagged with 'true /* at_return */? Same for line 240. The _x86_32 version uses false.

fisk replied:
Good point.

stefank commented on src/hotspot/share/c1/c1_Runtime1.cpp:
```diff
@@ -505,6 +506,21 @@ JRT_ENTRY_NO_ASYNC(static address, exception_handler_for_pc_helper(JavaThread* t
+  // stack frames are made safe before returning into them.
+  if (thread->last_frame().is_runtime_frame()) {
+    // The Runtime1::handle_exception_from_callee_id handler is invoked after the
+    // frame has been unwinded. It instead builds its own stub frame, to call the
```
Unwided -> unwound, maybe? Same below and probably other places as well.

fisk replied:
Good point.

stefank commented on src/hotspot/share/compiler/oopMap.cpp:
```diff
@@ -229,12 +235,17 @@ static void trace_codeblob_maps(const frame *fr, const RegisterMap *reg_map) {
 void OopMapSet::oops_do(const frame *fr, const RegisterMap* reg_map, OopClosure* f) {
   // add derived oops to a table
-  all_do(fr, reg_map, f, add_derived_oop, &do_nothing_cl);
+  if (!StackWatermarkSet::has_watermark(reg_map->thread(), StackWatermarkKind::gc) ||
+      DerivedPointerTable::is_active()) {
+    all_do(fr, reg_map, f, add_derived_oop, &do_nothing_cl);
+  } else {
+    all_do(fr, reg_map, f, process_derived_oop, &do_nothing_cl);
+  }
```
I wonder if we shouldn't hide the StackWatermarkSet in the GC code, and not "activate" the DerivedPointerTable when a SWS is used? Isn't it the case that already don't enable the table for ZGC? Couldn't this simply be:
`
  if (DerivedPointerTable::is_active()) {
    all_do(fr, reg_map, f, add_derived_oop, &do_nothing_cl);
  } else {
    all_do(fr, reg_map, f, process_derived_oop, &do_nothing_cl);
  }
`

fisk replied:
The problem isn't the GC code deciding to use or not use the derived pointer "table". It's the shared code uses of it that is the problem here. The table is explicitly activated by shared code, when for example using the JFR leak profiler, JVMTI heap walks, etc. This code makes the selection a GC choice when invoked by GC code, and a shared runtime choice when invoked by the shared runtime.

fisk replied:
After discussions off-line, I think passing in an explicit derived pointer iteration mode into oops_do, is much better than trying to figure out who the caller is from this context.

stefank commented on src/hotspot/share/compiler/oopMap.cpp:
```diff
@@ -297,8 +308,8 @@ void OopMapSet::all_do(const frame *fr, const RegisterMap *reg_map,
 #ifdef ASSERT
-        if ((((uintptr_t)loc & (sizeof(*loc)-1)) != 0) ||
-            !Universe::heap()->is_in_or_null(*loc)) {
+        if ((((uintptr_t)loc & (sizeof(oop)-1)) != 0) ||
+            !Universe::heap()->is_in_or_null((oop)NativeAccess<AS_NO_KEEPALIVE>::oop_load(&val))) {
```
Discussed offline. This is technical debt that we'd like to get rid of somehow.

stefank reviewed (commented):
Partially re-reviewed the changes.

dcubed-ojdk commented:
@fisk - You are missing testing information for this PR.

fisk commented:
> @fisk - You are missing testing information for this PR.

Sorry about that. I have performed testing with ZGC enabled from tier1-7 (multiple times), and standard testing tier1-5. I have also stress tested the change with gc-test-suite, with full verification on (there is a bunch of it), and the GC interval reduced to very small (GC all the time). Recent versions of this patch I also tested from tier1-8 (2 times). I have also perf tested this in dev-submit to check for regressions (none found), as well as stress tested SPECjbb2015 on a 3 TB machine with ~2200 threads. I have also manually stress tested the changes on an AArch64 machine, using gc-test-suite.

dcubed-ojdk commented:
@fisk - Thanks for the info. An amazing amount of testing!!

stefank commented on src/hotspot/share/gc/z/zRootsIterator.cpp:
```diff
@@ -268,10 +193,41 @@ void ZConcurrentRootsIterator::do_class_loader_data_graph(ZRootsIteratorClosure*
+private:
+  ZRootsIteratorClosure* const _cl;
+  // The resource mark is needed because interpreter oop maps are not reused in concurrent mode.
+  // Instead, they are temporary, and resource allocated.
```
temporary, and => temporary and ?

stefank commented on src/hotspot/share/gc/z/zStackWatermark.cpp:
```diff
@@ -0,0 +1,97 @@
+
+void ZStackWatermark::process(const frame& fr, RegisterMap& register_map, void* context) {
+  ZVerify::verify_frame_bad(fr, register_map);
+  frame(fr).oops_do(closure_from_context(context), &_cb_cl, &register_map);
```
With recent frame::oops_do cleanups, frame(fr) can be changed to simply fr.

stefank commented on src/hotspot/share/gc/z/zVerify.cpp:
```diff
@@ -206,3 +321,60 @@ ZVerifyViewsFlip::~ZVerifyViewsFlip() {
+
+void ZVerify::verify_frame_bad(const frame& fr, RegisterMap& register_map) {
+  ZVerifyBadOopClosure verify_cl;
+  frame(fr).oops_do(&verify_cl, NULL, &register_map);
```
With recent frame::oops_do cleanups, frame(fr) can be changed to simply fr.

stefank commented on src/hotspot/share/gc/z/zStackWatermark.hpp:
```diff
@@ -0,0 +1,68 @@
+public:
+  ZOnStackCodeBlobClosure();
+
+  virtual void do_code_blob(CodeBlob* cb);
```
Could be private.

stefank commented on src/hotspot/share/gc/z/zStackWatermark.hpp:
```diff
@@ -0,0 +1,68 @@
+
+  OopClosure* closure_from_context(void* context);
+
+protected:
```
No need to be protected, so can be private. We don't have sub-classes that overrides ZStackWatermark.

stefank reviewed (commented):
Next set of comments.

stefank commented on src/hotspot/share/runtime/stackWatermarkSet.cpp:
```diff
@@ -0,0 +1,145 @@
+#ifdef ASSERT
+  Thread* thread = Thread::current();
+  if (thread->is_Java_thread()) {
+    JavaThread* jt = static_cast<JavaThread*>(thread);
```
as_Java_thread()

stefank commented on src/hotspot/share/runtime/stackWatermarkSet.cpp:
```diff
@@ -0,0 +1,145 @@
+    // Don't perform barrier when error reporting walks the stack.
+    return;
+  }
+  verify_poll_context();
```
There's a verfy_poll_context here, but no update_poll_values call. I guess this is because we could be iterating over a thread that is not Thread::current()? But in that case, should we really be "verifying the poll context" of the current thread?

fisk replied:
Correct - I'm not updating the poll values, because it's not necessarily the current thread we are processing. But the verify_poll_context() code does correctly always refer to the current frame. It verifies we are in the right context to perform this kind of processing, yet the processing could be for a different thread.

fisk replied:
Updating the name to verify_processing_context to better reflect what this does.

stefank commented on src/hotspot/share/runtime/stackWatermarkSet.cpp:
```diff
@@ -0,0 +1,145 @@
+  if (watermark != NULL) {
+    watermark->start_processing();
+  }
+}
```
No update poll values here?

fisk replied:
Like the previous scenario, start_processing may be called on a different thread than Thread::current. The thread the stack belongs to will always update the poll values when it wakes up from the safepoint. Other threads have no business updating it.

stefank commented on src/hotspot/share/utilities/vmError.cpp:
```diff
@@ -749,8 +749,9 @@ void VMError::report(outputStream* st, bool _verbose) {
-       if (jt != NULL) {
+       Thread* thread = ((NamedThread *)_thread)->processed_thread();
+       if (thread != NULL && thread->is_Java_thread()) {
+         JavaThread* jt = static_cast<JavaThread*>(thread);
```
as_Java_thread() - maybe just search for this in the patch

fisk replied:
Done.

stefank reviewed (commented):
I've gone over the entire patch, but I'm leaving the compiler parts to others to review.

fisk commented on src/hotspot/share/gc/z/zStackWatermark.cpp:
```diff
@@ -0,0 +1,97 @@
+  _jt->oops_do_no_frames(closure_from_context(context), &_cb_cl);
+  ZThreadLocalData::do_invisible_root(_jt, ZBarrier::load_barrier_on_invisible_root_oop_field);
+
+  ZVerify::verify_thread_frames_bad(_jt);
```
I will try this.

fisk commented on src/hotspot/share/gc/z/zCollectedHeap.cpp:
```diff
@@ -229,6 +229,10 @@ bool ZCollectedHeap::can_elide_tlab_store_barriers() const {
   return false;
 }
 
+bool ZCollectedHeap::uses_stack_watermark_barrier() const {
+  return true;
+}
+
```
Perfect, thanks.

fisk commented on src/hotspot/share/gc/z/zDriver.cpp:
```diff
@@ -98,6 +102,10 @@ class VM_ZOperation : public VM_Operation {
     ZStatSample(ZSamplerJavaThreads, Threads::number_of_threads());
   }
 
+  virtual bool allow_coalesced_vm_operations() const {
+    return false;
+  }
+
```
Per explicitly wanted skip_thread_oop_barriers grouped with needs_inactive_gc_locker. But I should remove allow_coalesced_vm_operations; it is no longer used since I have removed VM operation coalescing completely already.

stefank replied:
I agree with Per, but I also wanted to move allow_coalesced_vm_operations. But now your removing it, so it becomes a moot point.

fisk commented on src/hotspot/share/jfr/leakprofiler/checkpoint/rootResolver.cpp:
```diff
@@ -272,7 +272,7 @@ bool ReferenceToThreadRootClosure::do_thread_stack_detailed(JavaThread* jt) {
 
     // Traverse the execution stack
-    for (StackFrameStream fst(jt); !fst.is_done(); fst.next()) {
+    for (StackFrameStream fst(jt, true /* update */, true /* process_frames */); !fst.is_done(); fst.next()) {
```
Will do.

fisk commented:
Thanks for reviewing @stefank, and thanks for all the pre-reviewing as well. I have updated the PR to resolve all your comments.

albertnetymk commented:
Given names like `StackWatermarkSet::lowest_watermark`, I wonder if some diagrams could be provided in the code comments for `class StackWatermarks` so that readers will have the correct mental pictures of the layout of stack frames and watermarks.

Some docs on `volatile uint32_t _state;` in `StackWatermark` would be nice, such as what info is encoded in this state, typical state transitions, who can access this state (tying it to volatile) etc.

I wonder if it's possible to remove `bool _is_done;` in `StackWatermarkFramesIterator`; just calling `is_done()` on the underlying frame stream seems cleaner.

fisk commented:
Thanks for reviewing Albert.

> Given names like `StackWatermarkSet::lowest_watermark`, I wonder if some diagrams could be provided in the code comments for `class StackWatermarks` so that readers will have the correct mental pictures of the layout of stack frames and watermarks.

Added some further comments and diagrams.
 
> Some docs on `volatile uint32_t _state;` in `StackWatermark` would be nice, such as what info is encoded in this state, typical state transitions, who can access this state (tying it to volatile) etc.

Added some comments about the state.

> I wonder if it's possible to remove `bool _is_done;` in `StackWatermarkFramesIterator`; just calling `is_done()` on the underlying frame stream seems cleaner.

Unfortunately this is not possible. The underlying frame stream is_done() function is not idempotent. It alters the state of the iteration as a side effect of asking the question whether it is done or not. The first time it reaches past the last frame it will report true, and the second time it will report false. So for that use to be valid, it is assumed that you only ask if you are is_done() once. But the way I use it, I ask this question potentially more than once after we are done, and expect it to return false consistently. That's why I read it once from the stream and cache the result that I use. One might argue that the underlying stream class has undesirable behaviour, but fixing that seemed outside the scope of this RFE. I have had too many rabbit holes already.

pliden commented on src/hotspot/share/gc/z/zBarrierSet.cpp:
```diff
@@ -89,6 +81,11 @@ void ZBarrierSet::on_thread_destroy(Thread* thread) {
   ZThreadLocalData::set_address_bad_mask(thread, ZAddressBadMask);
+  if (thread->is_Java_thread()) {
+    JavaThread* const jt = thread->as_Java_thread();
+    StackWatermark* watermark = new ZStackWatermark(jt);
```
Please add a const here.

pliden commented on src/hotspot/share/gc/z/zStackWatermark.cpp:
```diff
@@ -0,0 +1,95 @@
+void ZOnStackCodeBlobClosure::do_code_blob(CodeBlob* cb) {
+  nmethod* const nm = cb->as_nmethod_or_null();
+  if (nm != NULL) {
+    bool result = _bs_nm->nmethod_entry_barrier(nm);
```
Please add a const here.

pliden commented on src/hotspot/share/gc/z/zVerify.cpp:
```diff
@@ -55,16 +61,125 @@ static void z_verify_possibly_weak_oop(oop* p) {
+
+  virtual void do_thread(Thread* thread);
+
+  bool verify_fixed() const { return _verify_fixed; }
```
Please turn this into three lines.

pliden commented on src/hotspot/share/gc/z/zVerify.cpp:
```diff
@@ -55,16 +61,125 @@ static void z_verify_possibly_weak_oop(oop* p) {
+      _jt(jt),
+      _last_good(0),
+      _verifying_bad_frames(false) {
+    ZStackWatermark* stack_watermark = StackWatermarkSet::get<ZStackWatermark>(jt, StackWatermarkKind::gc);
```
Please add const here.

pliden commented on src/hotspot/share/gc/z/zVerify.cpp:
```diff
@@ -55,16 +61,125 @@ static void z_verify_possibly_weak_oop(oop* p) {
+    // The verification has two modes, depending on whether we have reached the
+    // last processed frame or not. Before it is reached, we expect everything to
+    // be good. After reaching it, we expect everything to be bad.
+    uintptr_t sp = reinterpret_cast<uintptr_t>(frame.sp());
```
Please add const here.

pliden reviewed (approved):
I had also pre-reviewed, but did a final semi-quick review. Looks good overall. Just found a couple of minor nitpicks.

robehn reviewed (approved):
Looks good.

fisk commented:
> I had also pre-reviewed, but did a final semi-quick review. Looks good overall. Just found a couple of minor nitpicks.

Thansk for the review Per. I uploaded fixes to your nitpicks.

fisk commented:
> Looks good.

Thanks for the review Robbin.

neliasso commented:
I did a pre-reviewed of the compiler parts, and have now done a second reading.

The compiler parts looks good!

fisk commented:
> I did a pre-reviewed of the compiler parts, and have now done a second reading.
> 
> The compiler parts looks good!

Thanks for the review, Nils.

coleenp commented on src/hotspot/share/runtime/mutexLocker.cpp:
```diff
@@ -246,7 +246,7 @@ void mutex_init() {
   def(Patching_lock                , PaddedMutex  , special,     true,  _safepoint_check_never);      // used for safepointing and code patching.
   def(CompiledMethod_lock          , PaddedMutex  , special-1,   true,  _safepoint_check_never);
-  def(Service_lock                 , PaddedMonitor, special,     true,  _safepoint_check_never);      // used for service thread operations
+  def(Service_lock                 , PaddedMonitor, tty-2,       true,  _safepoint_check_never);      // used for service thread operations
```
@pchilano Look.  tty-2

coleenp commented on src/hotspot/share/runtime/vframe.hpp:
```diff
@@ -337,7 +337,7 @@ class vframeStreamCommon : StackObj {
  public:
   // Constructors
-  vframeStream(JavaThread* thread, bool stop_at_java_call_stub = false);
+  vframeStream(JavaThread* thread, bool stop_at_java_call_stub = false, bool process_frames = true);
```
I was wondering if you supply arguments to all callers of vframeStream now?  It seems like having default parameters is an invitation to errors. Same with RegMap.

fisk replied:
I have tried to have explicit arguments as much as possible. But RegMap and vframeStream seemed to clutter my whole patch if I made them explicit (they are already implicit today). I agree though that having them explicit would be better. But what do you think about doing that in a follow-up cleanup RFE instead? It might reduce the amount of scrolling needed to review this patch. The defaults are on the safe conservative side, so not supplying the extra parameter should not harm you.

coleenp commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -2437,10 +2437,10 @@ void java_lang_Throwable::fill_in_stack_trace(Handle throwable, const methodHand
   // trace as utilizing vframe.
 #ifdef ASSERT
-  vframeStream st(thread);
+  vframeStream st(thread, false /* stop_at_java_call_stub */, false /* process_frames */);
```
This is a nit, but could you put the /* stop_at_java_call_stub*/ and /* process_frames */ before the values? Having a bunch of bool parameters seems like it could also become buggy.  In linkResolver.hpp we have enums so that the values can be checked and are more explicit at the call sites.  Not this change, but can we fix this later to do the same thing?

fisk replied:
Per and Stefan would kill me if I put the comments before the value (they explicitly prefer it on the right side of the value). Personally I have no strong preference. And yes, this also sounds like a good follow-up cleanup RFE.

coleenp reviewed (commented):
I looked at the runtime code and it is nicely non-invasive, and makes sense where there are changes.

fisk commented:
I uploaded an incremental change to re-instate the separate verification of the thread head vs frames that I had before. Stefan wanted it merged into one function unless I could remember why I had split it. I did not remember why, but now I do. Updated with comments this time to describe why this is split (the exception oop must be fixed before fiddling with the frames).

coleenp reviewed (approved):
To be clear, my comments about boolean parameters to vframeStream/RegMap can be addressed in a follow up RFE.  That would be better.

fisk commented:
> To be clear, my comments about boolean parameters to vframeStream/RegMap can be addressed in a follow up RFE.  That would be better.

Thanks for reviewing Coleen!

albertnetymk commented:
Thank you for the comments and diagrams; they make the code much more digestible. From that diagram, I get the impression that the watermark is associated with stack pointer, so it should be 1:1 relation, but `class Thread` contains multiple watermarks, `StackWatermarks _stack_watermarks;`. I think some high level description on the relation between the thread and a list of watermarks belong to that thread could be beneficial.

> The first time it reaches past the last frame it will report true, and the second time it will report false.

Why so? As I see it, once a stream becomes "done", it stays in that state forever. Am I missing sth?
```
inline bool StackFrameStream::is_done() {
  return (_is_done) ? true : (_is_done = _fr.is_first_frame(), false);
}
```

fisk commented:
> Thank you for the comments and diagrams; they make the code much more digestible. From that diagram, I get the impression that the watermark is associated with stack pointer, so it should be 1:1 relation, but `class Thread` contains multiple watermarks, `StackWatermarks _stack_watermarks;`. I think some high level description on the relation between the thread and a list of watermarks belong to that thread could be beneficial.

I added some further comments explaining this.

> > The first time it reaches past the last frame it will report true, and the second time it will report false.
> 
> Why so? As I see it, once a stream becomes "done", it stays in that state forever. Am I missing sth?
> 
> ```
> inline bool StackFrameStream::is_done() {
>   return (_is_done) ? true : (_is_done = _fr.is_first_frame(), false);
> }
> ```

When you step into the last frame of the iteration (first frame in the stack), the first time you ask is_done() it will report false, the next time it will report true, despite still being in the same frame. Therefore, the is_done() function is not idempotent, as I need it to be.

albertnetymk commented:
I see; thank you for the explanation.

rkennke commented on src/hotspot/share/compiler/oopMap.cpp:
```diff
@@ -297,8 +315,8 @@ void OopMapSet::all_do(const frame *fr, const RegisterMap *reg_map,
-        if ((((uintptr_t)loc & (sizeof(*loc)-1)) != 0) ||
-            !Universe::heap()->is_in_or_null(*loc)) {
+        if ((((uintptr_t)loc & (sizeof(oop)-1)) != 0) ||
+            !Universe::heap()->is_in_or_null((oop)NativeAccess<AS_NO_KEEPALIVE>::oop_load(&val))) {
```
This NativeAccess call causes the troubles with Shenandoah GC. It triggers evacuation of the object referenced by the stack location during a safepoint, which we'd rather avoid. Why is it needed?

fisk replied:
We call frame::oops_do from the GC to make bad oops good. But the oops_do logic assers the oop is good *before* the closure is applied.
Every time I do something, I run i to issues with this assert. It seems like an invalid assumption to me, that oops are meant to always be good, even before they are fixed. I vote for removing this assert. It has caused a lot of pain. Any takers?

shipilev replied:
I think the previous assert was intentionally weaker: `is_in` checks that argument points to a committed area in the heap, which can include the oops not yet fixed. It does not check if oop is valid as far as GC is concerned. So maybe reverting `NativeAccess::oop_load` is enough?

fisk replied:
It was checking is_in_or_null before and after. Our is_in checks that it is in committed memory, and that the pointer is not stale w.r.t. color, as that is very likely to be a bug. We reluctantly made it relaxed for our safepoints that flip colors. I think it was once again for this very same usual suspect assert. But now that this is concurrent, its memory can get concurrently freed.

I think this is the 7th time I try to make this assert happy. I run into it all the time. IMO, the underlying assumption of the assert, is wrong. This is an iterator used by GC to fix totally invalid stale pointers into valid pointers. You really can't expect that they are valid before fixing them. That just happened to be true for other GCs. It makes little sense to me. That is why my preferred solution is to remove the assert. I would not miss it.

We have had similar issues with the oop_iterate framework asserting oops are valid before the closure is applied. At least there, it is possible to opt out...

Would anyone miss the assert?

stefank replied:
FTR, this is the RFE to remove the oop verification from the oop_iterate framework:
https://bugs.openjdk.java.net/browse/JDK-8237363

I really would like to get rid of it, but got push back because it made GCs have to duplicate the effort to provide verification.

pliden replied:
I've also has problems with this assert in the past, and I think that the underlying assumption of the assert is wrong. I would not miss it.

rkennke replied:
IMO it's ok to remove it.
However, it can be argued that is_in() should not check for 'in committed memory' but 'in reserved space for heap', IOW 'is this a pointer into our heap memory region?'.  Or maybe this would be CH::is_in_reserved() and we should change the assert to that?
Also, IMO we should not do anything that may trip barriers on oop-iterators or similar GC-only paths.

fisk replied:
Almost always, what you want to do, is to assert the validity of heap pointers, where they are expected to be valid (dereferenceable pointer to a not freed object). And then you actually do want to check that they are within the committed boundaries of the heap. Otherwise you have a bug. We don't want to make such reasonable assertions weaker, because we cater for asserts that want oops to look valid pretty much all the time, even when they are stale pointers from a previous cycle, pointing into freed memory (until it is fixed). And I would much rather cater for asserts with good underlying reasoning (the oop should actually be okay here or something is wrong), than to cater for asserts that seem to just have the wrong underlying reasoning, making everyone else suffer for it, and letting bugs slip past all the other asserts.

The is_in_reserved check no longer exists. And I don't think I want to bring it back from the dead for this assert. Because I don't think the assert is right. The GCs already have their own verification code for roots (and the heap), that does this job, but much better. So I don't think it makes sense for the actual shared code oop iterator to go in and put constraints on what an invalid (not yet fixed up) stale pointer should and should not look like, for all GC implementations. Let the GC's own verification code take care of that. It has a much better idea about what a stale pointer should and should not look like.

And I do agree that we should not use barriers in this kind of code path. I just didn't know what else to do here to make the assert happy. Sorry about that. Well now I know. I am going to delete it, unless anyone opposes that solution.

rkennke replied:
It's ok for me.

zhengyu123 commented:
Hi Erik,

I have been playing with this patch for past a few days.  Great work!

I found that this patch seems to break an early assumption.
We have a comment in JavaThread::exit() says:

<pre>
  // We need to cache the thread name for logging purposes below as once
  // we have called on_thread_detach this thread must not access any oops.
</pre>

Then in method :
<pre><code>
void Threads::remove(JavaThread* p, bool is_daemon)  {
 ...
  BarrierSet::barrier_set()->on_thread_detach(p);

  // Extra scope needed for Thread_lock, so we can check
  // that we do not remove thread without safepoint code notice
  { MonitorLocker ml(Threads_lock);
.. 
}
</code></pre>
It calls barrier's on_thread_detach(), acquires Threads_lock.  
The lock acquisition triggers stack processing, that potential accesses oops.
<pre><code>
V  [libjvm.so+0x10c6f5e]  StackWatermark::start_processing()+0x6a
V  [libjvm.so+0x10c77e8]  StackWatermarkSet::start_processing(JavaThread*, StackWatermarkKind)+0x82
V  [libjvm.so+0xfd7757]  SafepointMechanism::process(JavaThread*)+0x37
V  [libjvm.so+0xfd796b]  SafepointMechanism::process_if_requested_slow(JavaThread*)+0x1d
V  [libjvm.so+0x4b3683]  SafepointMechanism::process_if_requested(JavaThread*)+0x2b
V  [libjvm.so+0xe87f0d]  ThreadBlockInVMWithDeadlockCheck::~ThreadBlockInVMWithDeadlockCheck()+0x5f
V  [libjvm.so+0xe86700]  Mutex::lock_contended(Thread*)+0x12c
V  [libjvm.so+0xe867d8]  Mutex::lock(Thread*)+0x96
V  [libjvm.so+0xe86823]  Mutex::lock()+0x23
V  [libjvm.so+0x29b4bc]  MutexLocker::MutexLocker(Mutex*, Mutex::SafepointCheckFlag)+0xe2
V  [libjvm.so+0x29b533]  MonitorLocker::MonitorLocker(Monitor*, Mutex::SafepointCheckFlag)+0x29
V  [libjvm.so+0x119f2ce]  Threads::remove(JavaThread*, bool)+0x56
V  [libjvm.so+0x1198a2b]  JavaThread::exit(bool, JavaThread::ExitType)+0x905
V  [libjvm.so+0x1197fde]  JavaThread::post_run()+0x22
V  [libjvm.so+0x1193eae]  Thread::call_run()+0x230
V  [libjvm.so+0xef3e38]  thread_native_entry(Thread*)+0x1e4
</code></pre>

This is a problem for Shenandoah, as it flushes SATB buffers during on_thread_detach() and does not expect to see any more SATB traffic.

Thanks.

fisk commented:
> Hi Erik,
> 
> I have been playing with this patch for past a few days. Great work!
> 
> I found that this patch seems to break an early assumption.
> We have a comment in JavaThread::exit() says:
> 
>   // We need to cache the thread name for logging purposes below as once
>   // we have called on_thread_detach this thread must not access any oops.
> 
> Then in method :
> 
> ```
> 
> void Threads::remove(JavaThread* p, bool is_daemon)  {
>  ...
>   BarrierSet::barrier_set()->on_thread_detach(p);
> 
>   // Extra scope needed for Thread_lock, so we can check
>   // that we do not remove thread without safepoint code notice
>   { MonitorLocker ml(Threads_lock);
> .. 
> }
> ```
> 
> It calls barrier's on_thread_detach(), acquires Threads_lock.
> The lock acquisition triggers stack processing, that potential accesses oops.
> 
> ```
> 
> V  [libjvm.so+0x10c6f5e]  StackWatermark::start_processing()+0x6a
> V  [libjvm.so+0x10c77e8]  StackWatermarkSet::start_processing(JavaThread*, StackWatermarkKind)+0x82
> V  [libjvm.so+0xfd7757]  SafepointMechanism::process(JavaThread*)+0x37
> V  [libjvm.so+0xfd796b]  SafepointMechanism::process_if_requested_slow(JavaThread*)+0x1d
> V  [libjvm.so+0x4b3683]  SafepointMechanism::process_if_requested(JavaThread*)+0x2b
> V  [libjvm.so+0xe87f0d]  ThreadBlockInVMWithDeadlockCheck::~ThreadBlockInVMWithDeadlockCheck()+0x5f
> V  [libjvm.so+0xe86700]  Mutex::lock_contended(Thread*)+0x12c
> V  [libjvm.so+0xe867d8]  Mutex::lock(Thread*)+0x96
> V  [libjvm.so+0xe86823]  Mutex::lock()+0x23
> V  [libjvm.so+0x29b4bc]  MutexLocker::MutexLocker(Mutex*, Mutex::SafepointCheckFlag)+0xe2
> V  [libjvm.so+0x29b533]  MonitorLocker::MonitorLocker(Monitor*, Mutex::SafepointCheckFlag)+0x29
> V  [libjvm.so+0x119f2ce]  Threads::remove(JavaThread*, bool)+0x56
> V  [libjvm.so+0x1198a2b]  JavaThread::exit(bool, JavaThread::ExitType)+0x905
> V  [libjvm.so+0x1197fde]  JavaThread::post_run()+0x22
> V  [libjvm.so+0x1193eae]  Thread::call_run()+0x230
> V  [libjvm.so+0xef3e38]  thread_native_entry(Thread*)+0x1e4
> ```
> 
> This is a problem for Shenandoah, as it flushes SATB buffers during on_thread_detach() and does not expect to see any more SATB traffic.
> 
> Thanks.

What oop are you encountering here? You should have no frames left at this point, and all oops should have been cleared. At least that is the theory, and that was why the thread oop moved out from the thread (to enforce that). So I am curious what oop you have found to still be around at this point.

Anyway, you can try moving the GC hook into the critical section. That should help.

zhengyu123 commented:
> > Hi Erik,
> > I have been playing with this patch for past a few days. Great work!
> > I found that this patch seems to break an early assumption.
> > We have a comment in JavaThread::exit() says:
> > // We need to cache the thread name for logging purposes below as once
> > // we have called on_thread_detach this thread must not access any oops.
> > Then in method :
> > ```
> > 
> > void Threads::remove(JavaThread* p, bool is_daemon)  {
> >  ...
> >   BarrierSet::barrier_set()->on_thread_detach(p);
> > 
> >   // Extra scope needed for Thread_lock, so we can check
> >   // that we do not remove thread without safepoint code notice
> >   { MonitorLocker ml(Threads_lock);
> > .. 
> > }
> > ```
> > 
> > 
> > It calls barrier's on_thread_detach(), acquires Threads_lock.
> > The lock acquisition triggers stack processing, that potential accesses oops.
> > ```
> > 
> > V  [libjvm.so+0x10c6f5e]  StackWatermark::start_processing()+0x6a
> > V  [libjvm.so+0x10c77e8]  StackWatermarkSet::start_processing(JavaThread*, StackWatermarkKind)+0x82
> > V  [libjvm.so+0xfd7757]  SafepointMechanism::process(JavaThread*)+0x37
> > V  [libjvm.so+0xfd796b]  SafepointMechanism::process_if_requested_slow(JavaThread*)+0x1d
> > V  [libjvm.so+0x4b3683]  SafepointMechanism::process_if_requested(JavaThread*)+0x2b
> > V  [libjvm.so+0xe87f0d]  ThreadBlockInVMWithDeadlockCheck::~ThreadBlockInVMWithDeadlockCheck()+0x5f
> > V  [libjvm.so+0xe86700]  Mutex::lock_contended(Thread*)+0x12c
> > V  [libjvm.so+0xe867d8]  Mutex::lock(Thread*)+0x96
> > V  [libjvm.so+0xe86823]  Mutex::lock()+0x23
> > V  [libjvm.so+0x29b4bc]  MutexLocker::MutexLocker(Mutex*, Mutex::SafepointCheckFlag)+0xe2
> > V  [libjvm.so+0x29b533]  MonitorLocker::MonitorLocker(Monitor*, Mutex::SafepointCheckFlag)+0x29
> > V  [libjvm.so+0x119f2ce]  Threads::remove(JavaThread*, bool)+0x56
> > V  [libjvm.so+0x1198a2b]  JavaThread::exit(bool, JavaThread::ExitType)+0x905
> > V  [libjvm.so+0x1197fde]  JavaThread::post_run()+0x22
> > V  [libjvm.so+0x1193eae]  Thread::call_run()+0x230
> > V  [libjvm.so+0xef3e38]  thread_native_entry(Thread*)+0x1e4
> > ```
> > 
> > 
> > This is a problem for Shenandoah, as it flushes SATB buffers during on_thread_detach() and does not expect to see any more SATB traffic.
> > Thanks.
> 
> What oop are you encountering here? You should have no frames left at this point, and all oops should have been cleared. At least that is the theory, and that was why the thread oop moved out from the thread (to enforce that). So I am curious what oop you have found to still be around at this point.

They are handles:
<pre><code>
V  [libjvm.so+0x98d693]  chunk_oops_do(OopClosure*, Chunk*, char*)+0xce
V  [libjvm.so+0x98d6d3]  HandleArea::oops_do(OopClosure*)+0x37
V  [libjvm.so+0x1194ad0]  Thread::oops_do_no_frames(OopClosure*, CodeBlobClosure*)+0x88
V  [libjvm.so+0x119b0c8]  JavaThread::oops_do_no_frames(OopClosure*, CodeBlobClosure*)+0x88
</code></pre>

> 
> Anyway, you can try moving the GC hook into the critical section. That should help.

This seems to work.

fisk commented:
> > > Hi Erik,
> > > I have been playing with this patch for past a few days. Great work!
> > > I found that this patch seems to break an early assumption.
> > > We have a comment in JavaThread::exit() says:
> > > // We need to cache the thread name for logging purposes below as once
> > > // we have called on_thread_detach this thread must not access any oops.
> > > Then in method :
> > > ```
> > > 
> > > void Threads::remove(JavaThread* p, bool is_daemon)  {
> > >  ...
> > >   BarrierSet::barrier_set()->on_thread_detach(p);
> > > 
> > >   // Extra scope needed for Thread_lock, so we can check
> > >   // that we do not remove thread without safepoint code notice
> > >   { MonitorLocker ml(Threads_lock);
> > > .. 
> > > }
> > > ```
> > > 
> > > 
> > > It calls barrier's on_thread_detach(), acquires Threads_lock.
> > > The lock acquisition triggers stack processing, that potential accesses oops.
> > > ```
> > > 
> > > V  [libjvm.so+0x10c6f5e]  StackWatermark::start_processing()+0x6a
> > > V  [libjvm.so+0x10c77e8]  StackWatermarkSet::start_processing(JavaThread*, StackWatermarkKind)+0x82
> > > V  [libjvm.so+0xfd7757]  SafepointMechanism::process(JavaThread*)+0x37
> > > V  [libjvm.so+0xfd796b]  SafepointMechanism::process_if_requested_slow(JavaThread*)+0x1d
> > > V  [libjvm.so+0x4b3683]  SafepointMechanism::process_if_requested(JavaThread*)+0x2b
> > > V  [libjvm.so+0xe87f0d]  ThreadBlockInVMWithDeadlockCheck::~ThreadBlockInVMWithDeadlockCheck()+0x5f
> > > V  [libjvm.so+0xe86700]  Mutex::lock_contended(Thread*)+0x12c
> > > V  [libjvm.so+0xe867d8]  Mutex::lock(Thread*)+0x96
> > > V  [libjvm.so+0xe86823]  Mutex::lock()+0x23
> > > V  [libjvm.so+0x29b4bc]  MutexLocker::MutexLocker(Mutex*, Mutex::SafepointCheckFlag)+0xe2
> > > V  [libjvm.so+0x29b533]  MonitorLocker::MonitorLocker(Monitor*, Mutex::SafepointCheckFlag)+0x29
> > > V  [libjvm.so+0x119f2ce]  Threads::remove(JavaThread*, bool)+0x56
> > > V  [libjvm.so+0x1198a2b]  JavaThread::exit(bool, JavaThread::ExitType)+0x905
> > > V  [libjvm.so+0x1197fde]  JavaThread::post_run()+0x22
> > > V  [libjvm.so+0x1193eae]  Thread::call_run()+0x230
> > > V  [libjvm.so+0xef3e38]  thread_native_entry(Thread*)+0x1e4
> > > ```
> > > 
> > > 
> > > This is a problem for Shenandoah, as it flushes SATB buffers during on_thread_detach() and does not expect to see any more SATB traffic.
> > > Thanks.
> > 
> > 
> > What oop are you encountering here? You should have no frames left at this point, and all oops should have been cleared. At least that is the theory, and that was why the thread oop moved out from the thread (to enforce that). So I am curious what oop you have found to still be around at this point.
> 
> They are handles:
> 
> ```
> 
> V  [libjvm.so+0x98d693]  chunk_oops_do(OopClosure*, Chunk*, char*)+0xce
> V  [libjvm.so+0x98d6d3]  HandleArea::oops_do(OopClosure*)+0x37
> V  [libjvm.so+0x1194ad0]  Thread::oops_do_no_frames(OopClosure*, CodeBlobClosure*)+0x88
> V  [libjvm.so+0x119b0c8]  JavaThread::oops_do_no_frames(OopClosure*, CodeBlobClosure*)+0x88
> ```
> 
> > Anyway, you can try moving the GC hook into the critical section. That should help.
> 
> This seems to work.

Okay, I will move it.

fisk commented:
> _Mailing list message from [Kim Barrett](mailto:kim.barrett@oracle.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> I've only looked at scattered pieces, but what I've looked at seemed to be
> in good shape. Only a few minor comments.
> 
> ------------------------------------------------------------------------------
> src/hotspot/share/runtime/frame.cpp
> 456 // for(StackFrameStream fst(thread); !fst.is_done(); fst.next()) {
> 
> Needs to be updated for the new constructor arguments. Just in general, the
> class documentation seems to need some updating for this change.

Fixed.

> ------------------------------------------------------------------------------
> src/hotspot/share/runtime/frame.cpp
> 466 StackFrameStream(JavaThread *thread, bool update, bool process_frames);
> 
> Something to consider is that bool parameters like that, especially when
> there are multiple, are error prone. An alternative is distinct enums, which
> likely also obviates the need for comments in calls.

Coleen also had the same comment, and we agreed to file a follow-up RFE to clean that up. This also applies to all the existing parameters passed in. So I would still like to do that, but in a follow-up RFE. Said RFE will also make the parameter use in RegisterMap and vframeStream explicit.

> ------------------------------------------------------------------------------
> src/hotspot/share/runtime/thread.hpp
> 956 void set_processed_thread(Thread *thread) { _processed_thread = thread; }
> 
> I think this should assert that either _processed_thread or thread are NULL.
> Or maybe the RememberProcessedThread constructor should be asserting that
> _cur_thr->processed_thread() is NULL.

Fixed.

> ------------------------------------------------------------------------------

Thanks for the review Kim!

fisk commented:
In my last PR update, I included a fix to an exception handling problem that I encountered after lots of stress testing that I have been running for a while now. I managed to catch the issue, get a reliable reproducer, and fix it.

The root problem is that the hook I had placed in SharedRuntime::exception_handler_for_return_address has been ignored. The reason is that the stack is not walkable at this point. The hook then just ignores it. This had some unexpected consequences. After looking closer at this code, I found that if we did have a walkable stack when we call SharedRuntime::raw_exception_handler_for_return_address, that would have been the only hook we need at all for exception handling. It is always the common root point where we unwind into a caller frame due to an exception throwing into the caller, and we need to look up the rethrow handler of the caller. However, we are indeed not walkable here. To deal with this, I have rearranged the exceptino hooks a bit. First of all, I have deleted all before_unwind hooks for exception handling, because they should not be needed if the after_unwind hook is reliably called on the caller side instead. And those hooks do indeed need to be there, because we do not always have a point where we can call before_unwind (e.g. C1 unwind exception code, that just unwinds and looks up the rethrow handler via SharedRuntime::exception_handler_for_return_address). I have then traced all paths from SharedRuntime::raw_exception_handler_for_return_address into runtime rethrow handlers called, for each rethrow exception handler PC exposed in the function. They are: 

* OptoRuntime::rethrow_C when unwinding into C2 code
* exception_handler_for_pc_helper via Runtime1::handle_exception_from_callee_id when unwinding into C1 code
* JavaCallWrapper::~JavaCallWrapper when unwinding into a Java call stub
* InterpreterRuntime::exception_handler_for_exception when unwinding into an interpreted method
* Deoptimization::fetch_unroll_info (with exec_mode == Unpack_exception) when unwinding into a deoptimized nmethod

Each rethrow handler returned has a corresponding comment saying which rethrow runtime rethrow handler it will end up in, once the stack has been walkable and we have transferred control into the caller. And all of those runtime hooks now have an after_unwind() hook.

The good news is that now the responsibility for who calls the unwind hook for exception is clearer: it is never done by the callee, and always done by the caller, in its rethrow handler, at which point the stack is walkable.

In order to avoid further issues where an unwind hook is ignored, I have changed them to assert that there is a last_Java_frame present. Previously I did not assert that, because there was shared code between runtime native transitions and the native wrapper, that called an unwind hook. This prevented the unwind hooks to assert this, as compiler threads would still perform native transitions, that just ignored the request. I moved the problematic hook for native up one level in the hierarchy to a path where it is only called by native wrappers (where we always have a last_Java_frame), so that I can finally assert that the unwind hooks always are called at points where we have a last_Java_frame. This makes me feel confident that I do not have another hook that is being accidentally ignored.

However, the relationship for the various exception handling code executed in the caller, the callee, and between the two (before we are walkable) is rather complicated. So it would be good to have someone that knows the exception code very well have a look at this, to make sure I have not missed anything. I have rerun all testing and done a load of stress testing to sanity check this. The reproducer I eventually found that reproduced the issue with 100% success rate, was run many times with the new patch, and no longer reproduces any issue.

dholmes-ora commented on src/hotspot/share/runtime/safepointMechanism.cpp:
```diff
@@ -78,29 +82,59 @@ void SafepointMechanism::process(JavaThread *thread) {
+  // The call to start_processing fixes the thread's oops and the first few frames.
+  //
+  // The call has been carefully placed here to cater for a few situations:
+  // 1) After we exit from block after a global pool
```
Typo: pool -> poll

fisk replied:
Fixed.

dholmes-ora commented on src/hotspot/share/runtime/stackWatermark.cpp:
```diff
@@ -0,0 +1,301 @@
+
+void StackWatermark::yield_processing() {
+  update_watermark();
+  MutexUnlocker mul(&_lock, Mutex::_no_safepoint_check_flag);
```
This seems a little dubious - is it just a heuristic? There is no guarantee that unlocking the Mutex will allow another thread to claim it before this thread re-locks it.

fisk replied:
It is indeed just a heuristic. There is no need for a guarantee.

dholmes-ora commented on src/hotspot/share/runtime/stackWatermark.hpp:
```diff
@@ -0,0 +1,144 @@
+  StackWatermark* _next;
+  JavaThread* _jt;
+  StackWatermarkFramesIterator* _iterator;
+  Mutex _lock;
```
How are you guaranteeing that the Mutex is unused at the time the StackWatermark is deleted?

fisk replied:
The StackWatermarks are deleted when the thread is deleted (and its destructor runs). Hence, I'm relying on the Threads SMR project here. Anyone that pokes around at the StackWatermark is either the current thread, or a thread that has a ThreadsListHandle containing the thread, making it safe to access that thread without it racingly being deleted.

dholmes-ora commented on src/hotspot/share/runtime/safepointMechanism.cpp:
```diff
@@ -78,29 +82,59 @@ void SafepointMechanism::process(JavaThread *thread) {
 
+uintptr_t SafepointMechanism::compute_poll_word(bool armed, uintptr_t stack_watermark) {
+  if (armed) {
+    log_debug(stackbarrier)("Computed armed at %d", Thread::current()->osthread()->thread_id());
```
s/at/for/ ?

fisk replied:
Fixed.

dholmes-ora commented:
Hi Erik,
Can you give an overview of the use of the "poll word" and its relation to the "poll page" please?
Thanks,
David

fisk commented:
> Hi Erik,
> Can you give an overview of the use of the "poll word" and its relation to the "poll page" please?
> Thanks,
> David

Hi David,

Thanks for reviewing this code.

There are various polls in the VM. We have runtime transitions, interpreter transitions, transitions at returns, native wrappers, transitions in nmethods... and sometimes they are a bit different.

The "poll word" encapsulates enough information to be able to poll for returns (stack watermark barrier), or poll for normal handshakes/safepoints, with a conditional branch. So really, we could use the "poll word" for every single poll. A low order bit is a boolean saying if handshake/safepoint is armed, and the rest of the word denotes the watermark for which frame has armed returns.

The "poll page" is for polls that do not use conditional branches, but instead uses an indirect load. It is used still in nmethod loop polls, because I experimentally found it to perform worse with conditional branches on one machine, and did not want to risk regressions. It is also used for VM configurations that do not yet support stack watermark barriers, such as Graal, PPC, S390 and 32 bit platforms. They will hopefully eventually support this mechanism, but having the poll page allows a more smooth transition. And unless it is crystal clear that the performance of the conditional branch loop poll really is fast enough on sufficiently many machines, we might keep it until that changes.

Hope this makes sense.

Thanks,
/Erik

fisk commented:
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> On 06/10/2020 08:22, Erik ?sterlund wrote:
> 
> > > This PR the implementation of "JEP 376: ZGC: Concurrent Thread-Stack Processing" (cf.
> > > https://openjdk.java.net/jeps/376).
> 
> One small thing: the couple of uses of lea(InternalAddress) should really be adr;
> this generates much better code.

Hi Andrew,

Thanks for having a look. I applied your patch. Having said that, this is run on the safepoint slow path, so should be a rather cold path, where threads have to wear coats and gloves. But it does not hurt to optimize the encoding further, I suppose.

Thanks,
/Erik

fisk commented:
> <!-- Bridged id (YjQ0ZjU5NTMtZDViNC1kYmQ4LTlhYTMtMGI2N2IwMzU0YzQ4QG9yYWNsZS5jb20=) -->
> *Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [serviceability-dev](mailto:serviceability-dev@openjdk.java.net):*
> 
> Hi Erik\,
> 
> On 6\/10\/2020 5\:37 pm\, Erik \?sterlund wrote\:
> > On Tue\, 6 Oct 2020 02\:57\:00 GMT\, David Holmes \<dholmes at openjdk\.org> wrote\:
> > 
> >> Hi Erik\,
> >> Can you give an overview of the use of the \"poll word\" and its relation to the \"poll page\" please\?
> >> Thanks\,
> >> David
> > 
> > Hi David\,
> > 
> > Thanks for reviewing this code\.
> > 
> > There are various polls in the VM\. We have runtime transitions\, interpreter transitions\, transitions at returns\, native
> > wrappers\, transitions in nmethods\.\.\. and sometimes they are a bit different\.
> > 
> > The \"poll word\" encapsulates enough information to be able to poll for returns \(stack watermark barrier\)\, or poll for
> > normal handshakes\/safepoints\, with a conditional branch\. So really\, we could use the \"poll word\" for every single poll\.
> > A low order bit is a boolean saying if handshake\/safepoint is armed\, and the rest of the word denotes the watermark for
> > which frame has armed returns\.
> > 
> > The \"poll page\" is for polls that do not use conditional branches\, but instead uses an indirect load\. It is used still
> > in nmethod loop polls\, because I experimentally found it to perform worse with conditional branches on one machine\, and
> > did not want to risk regressions\. It is also used for VM configurations that do not yet support stack watermark
> > barriers\, such as Graal\, PPC\, S390 and 32 bit platforms\. They will hopefully eventually support this mechanism\, but
> > having the poll page allows a more smooth transition\. And unless it is crystal clear that the performance of the
> > conditional branch loop poll really is fast enough on sufficiently many machines\, we might keep it until that changes\.
> > 
> > Hope this makes sense\.
> 
> Yes but I am somewhat surprised\. The conventional wisdom has always been 
> that polling based on the \"poison page\" approach far outperforms 
> explicit load\-test\-branch approaches\.
> 
> Cheers\,
> David

When thread local handshakes was built, both a branch based and indirect load based prototype was implemented. I had a branch based solution and Mikael Gerdin built an indirect load based solution, so we could compare them. He compared them on many machines and found that sometimes branches are a bit faster and sometimes a bit slower, depending on CPU model. But the results with indirect loads was more stable from machine to machine, while the branch based solution depended a bit more on what CPU model was being used. That is why the indirect load solution was chosen: it was not always best but it was never bad on any machine.

Since then, we got loop strip mining in C2 which makes the frequency of polls less tight. My hypothesis was that with that in place, a new evaluation would show that branching is fine now. However, one machine did not agree with that. To be fair, that machine is not giving very stable results at all right now, so I am not sure if there is a real problem or not. But I thought I'll keep the old behaviour for now anyway as I do not yet have a good reason to change it, and not good enough proof that it is okay... yet.

/Erik

stooart-mon reviewed (approved):
I've been reviewing this and stepping through the debugger. It looks OK to me.

fisk commented:
> I've been reviewing this and stepping through the debugger. It looks OK to me.

Thanks for the review Stuart.

fisk commented:
/integrate
